tests/Validation.hs

```diff
@@ -0,0 +1,53 @@
+module Validation where
+
+import Control.Monad.Trans.Except
+import Crypto.PubKey.RSA (PublicKey)
+import qualified Data.ByteString as B
+import qualified Data.ByteString.Base64 as Base64
+import Data.Time.Format.ISO8601
+import qualified Data.X509 as X509
+import qualified Data.X509.Memory as X509
+import Network.Wai.SAML2
+import Network.Wai.SAML2.Validation
+import System.FilePath
+import Test.Tasty
+import Test.Tasty.ExpectedFailure
+import Test.Tasty.HUnit
+
+-- | Get a public key from a X.509 certificate
+parseCertificate :: B.ByteString -> PublicKey
+parseCertificate certificate = case X509.readSignedObjectFromMemory certificate of
+    [signedCert] -> case X509.certPubKey $ X509.signedObject $ X509.getSigned signedCert of
+        X509.PubKeyRSA key -> key
+        other -> error $ "Expected PubKeyRSA, but got " <> show other
+    xs -> error $ show xs
+
+run :: FilePath -> String -> FilePath -> IO ()
+run certPath timestamp respPath = do
+    cert <- B.readFile $ prefix </> certPath
+    xml <- B.readFile $ prefix </> respPath
+    now <- iso8601ParseM timestamp
+
+    let pub = parseCertificate cert
+        cfg = saml2ConfigNoEncryption pub
+
+    assertion <- runExceptT $ do
+        (responseXmlDoc, samlResponse) <- decodeResponse $ Base64.encode xml
+        validateSAMLResponse cfg responseXmlDoc samlResponse now
+
+    case assertion of
+        Left err -> assertFailure $ show err
+        Right _ -> pure ()
+
+prefix :: FilePath
+prefix = "tests/data"
+
+tests :: TestTree
+tests = testGroup "Validate SAML2 Response"
+    [ testCase "AzureAD signed response"
+        $ run "azuread.crt" "2023-05-10T01:20:00Z" "azuread-signed-response.xml"
+    , expectFail $ testCase "AzureAD signed assertion"
+        $ run "azuread.crt" "2023-05-09T16:00:00Z" "azuread-signed-assertion.xml"
+    , testCase "Okta with AttributeStatement"
+        $ run "okta.crt" "2023-06-16T06:43:00.000Z" "okta-attributes.xml"
```

Comment:
Before your most recent force push, this test was also guarded by `expectFail`:

```suggestion
    , expectFail $ testCase "Okta with AttributeStatement"
        $ run "okta.crt" "2023-06-16T06:43:00.000Z" "okta-attributes.xml"
```

I assume that this is why the CI fails for LTS 19 and older. The real question is why it _succeeds_ for LTS 20 when that should fail as well!

Reply:
This is intentional because #51 is now merged and I rebased the branch on master (otherwise it would fail after merging),

Reply:
I commented out this test case for now

Reply:
Ah I see, I hadn't realised that one of these test cases was for https://github.com/mbg/wai-saml2/pull/51. (See also my assumption that both tests were for https://github.com/mbg/wai-saml2/pull/45 in https://github.com/mbg/wai-saml2/pull/52#discussion_r1240718082).

It would be good for us to better understand why the test fails unexpectedly with the older Stack resolvers to at least establish whether this is an issue in one of the libraries we are using or some problem in our code. 

I am going to be on holiday from tomorrow for the next two weeks, so I don't have time right now to look more deeply into this, but I can do so when I am back if you want.

Reply:
@mbg I updated `stack.yaml`s to use the latest c14n. That did the trick